DOC: Add window transform example

iaindillingham opened the pull request:
This example shows that aggregation functions aren't only for aggregate transforms. By using two aggregation functions within a window transform we are able to compute z-scores within a calculate transform.

See also: https://vega-js.slack.com/archives/C0NG6SXFX/p1548451285140300

jakevdp commented on doc/user_guide/transform.rst:
```diff
@@ -699,6 +699,38 @@ last_value    None       Assigns a value from the last data object in the curren
+Notice that when we use an aggregation function within an aggregate transform, the data objects are collapsed.
+However, when we use an aggregation function within a window transform, the data objects are augmented.
+It's by augmenting, rather than collapsing, that we can compute statistics, such as `z-scores`_.
+For example, consider the following time series of stock prices:
```
I'm not entirely clear what is meant by "collapsed" and "augmented" in this context.

iaindillingham replied:
Yes, good question. What I want to communicate is that an aggregate transform with an aggregation function is like an SQL `GROUP BY` but a window transform with an aggregation function is like broadcasting in Pandas. In the latter case, each data object has a new property (is augmented with), however, there are the same number of data objects. In the former case, there are (probably) a different number of objects.

jakevdp replied:
Ah, maybe something like "While aggregate transforms compute a single value summarizing the dataset, window transforms effectively add a new column to the dataset, constructed from local aggregates at each row."

iaindillingham replied:
I'm not sure what you mean by "local aggregates around each value in the column". Local to the data object, which is synonymous with adding a new column? I updated this section in a previous commit, but Github's UI unfortunately isn't displaying it here. Would you mind taking a look? I'll comment it now.

jakevdp replied:
By "local aggregates" I was trying to say "windowed aggregates" without being so tautological :smile:

jakevdp replied:
It's local in the sense that the values nearby the point are aggregated and put in the new column.

iaindillingham replied:
Ah, I see! I've replaced the section with the following:

> While an aggregate transform computes a single value that summarises all data objects, a window transform adds a new property to each data object. This new property is computed from the neighbouring data objects: that is, from the data objects delimited by the window field definition.

I felt that "neighboring" was more descriptive than "local", but I'm happy to change it. Hopefully the reference to the window field definition isn't too tautological.

jakevdp replied:
Looks good, thanks!

jakevdp commented on doc/user_guide/transform.rst:
```diff
@@ -699,6 +699,38 @@ last_value    None       Assigns a value from the last data object in the curren
+    import altair as alt
+    from vega_datasets import data
+
+    alt.Chart(data.stocks.url).mark_line().encode(x='date:T', y='price:Q', color='symbol:N')
```
Please put each encoding argument indented on its own line.

jakevdp commented on doc/user_guide/transform.rst:
```diff
@@ -699,6 +699,38 @@ last_value    None       Assigns a value from the last data object in the curren
+        stdev_price='stdev(price)',
+        frame=[None, None], groupby=['symbol'],
+    ).transform_calculate(
+        z_score=(expr.datum.price - expr.datum.mean_price) / expr.datum.stdev_price,
```
Use ``alt.datum`` rather than ``expr.datum``

jakevdp commented on doc/user_guide/transform.rst:
```diff
@@ -699,6 +699,38 @@ last_value    None       Assigns a value from the last data object in the curren
+        frame=[None, None], groupby=['symbol'],
+    ).transform_calculate(
+        z_score=(expr.datum.price - expr.datum.mean_price) / expr.datum.stdev_price,
+    ).mark_line().encode(x='date:T', y='z_score:Q', color='symbol:N')
```
encodings on their own line, as above

iaindillingham commented:
I've addressed your comment about encoding arguments, and extended it slightly for consistency. I've attempted to clarify the aggregate/window difference with more descriptive concepts of `GROUP BY` and broadcasting, although couldn't find a definition of the latter in the Pandas documentation. Hopefully the combination of text and example is sufficient: it's the `GROUP BY`/broadcasting behaviour that initially confused me about aggregate/window transforms.

iaindillingham commented on doc/user_guide/transform.rst:
```diff
@@ -702,9 +702,10 @@ last_value    None       Assigns a value from the last data object in the curren
+the input data objects are replaced by the output data objects, which contain the results of applying the aggregation function.
+However, using a window transform is like broadcasting the results to the input data objects.
+By using a window transform, and preserving the input data objects, we can compute statistics, such as `z-scores`_.
 For example, consider the following time series of stock prices:
```
Does this address your original comment @jakevdp?

jakevdp replied:
I don't really understand the connection here between window functions and broadcasting. To me they seem like distinct concepts: for example, you cannot compute a convolution through broadcasting except via a very contrived process, but you can easily express a convolution via a window transform. I like your description elsewhere of window transforms as aggregates computed in the neighborhood of each row of the data.

Regarding group by & aggregation, I think this is also a confusing comparison. An altair aggregate can express the same thing as a groupby followed by an aggregation (if you specify the ``groupby`` parameter in the transform) but in general aggregation is not equivalent to a group-by.

iaindillingham replied:
I think we're out of step with our comments and commits. I've removed the references to `GROUP BY` and broadcasting, as well as the original references to collapsed and augmented. My most recent commit incorporated your suggested amendments.

iaindillingham replied:
I've managed to get into a tangle with PRs and MRs on GitHub before. Sorry for the confusion!

jakevdp replied:
Ah, I see. Sorry about that!

jakevdp commented on doc/user_guide/transform.rst:
```diff
@@ -699,6 +702,46 @@ last_value    None       Assigns a value from the last data object in the curren
+.. altair-plot::
+
+    import altair as alt
+    from altair import expr
```
Import is no longer used

iaindillingham replied:
Good spot!

iaindillingham replied:
Removed in latest commit.

iaindillingham commented:
Okay, I think we're there! Thanks for your feedback, and, as every for your hard work on Altair.

jakevdp commented:
Thanks!
